chain/chain/src/state_transition_data.rs

```diff
@@ -0,0 +1,110 @@
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_chain_primitives::error::Error;
+use near_primitives::block::Block;
+use near_primitives::hash::CryptoHash;
+use near_primitives::types::{BlockHeight, ShardId};
+use near_primitives::utils::{get_block_shard_id, get_block_shard_id_rev};
+use near_store::db::STATE_TRANSITIONS_METADATA_KEY;
+use near_store::{DBCol, StorageError};
+use once_cell::unsync::OnceCell;
+
+use crate::{ChainStore, ChainStoreAccess};
+
+#[derive(BorshSerialize, BorshDeserialize, Default, Debug)]
+struct StateTransitionsMetadata {
+    /// Represents max not-yet-GCed height for each shard
+    start_heights: HashMap<ShardId, BlockHeight>,
+}
+
+/// Removes StateTransitionData entries from the db based on `block`.
+/// It is safe to delete all StateTransitionData before the last
+/// final block for shards with present chunks in that block.
+pub(crate) fn garbage_collect_state_transition_data(
+    chain_store: &ChainStore,
+    block: &Block,
+) -> Result<(), Error> {
+    let final_block_hash = *block.header().last_final_block();
+    if final_block_hash == CryptoHash::default() {
+        return Ok(());
+    }
+    let final_block = chain_store.get_block(&final_block_hash)?;
+    let final_block_height = final_block.header().height();
+    let mut metadata = chain_store
+        .store()
+        .get_ser::<StateTransitionsMetadata>(DBCol::Misc, STATE_TRANSITIONS_METADATA_KEY)?
+        .unwrap_or_default();
+    tracing::debug!(
+        target: "state_transition_data",
+        final_block_height,
+        ?metadata,
+        "garbage collecting state transition data"
+    );
+    // Only calculated if metadata doesn't exist or if it doesn't contain start_heights entry for the shard
+    let computed_start_heights = OnceCell::<HashMap<ShardId, BlockHeight>>::new();
+    let mut store_update = chain_store.store().store_update();
+    for chunk in final_block.chunks().iter().filter(|chunk| chunk.is_new_chunk(final_block_height))
+    {
+        let shard_id = chunk.shard_id();
+        let start_height = if let Some(&start_height) = metadata.start_heights.get(&shard_id) {
+            start_height
+        } else {
+            *computed_start_heights
+                .get_or_try_init(|| compute_start_heights(chain_store))?
+                .get(&shard_id)
+                .unwrap_or(&final_block_height)
+        };
+        let mut deleted_count = 0;
+        for height in start_height..final_block_height {
```

Comment:
Is it legal to remove all state transitions before the final block?
When generating a state witness for a chunk, we look for state transition data for the block that included the previous chunk. If there are a lot of missing chunks on some shard, this block could maybe end up being older than the final block.
https://github.com/near/nearcore/blob/61cec49e903472d4b45d3131c570aaff811358d3/chain/client/src/stateless_validation/state_witness_producer.rs#L149

Reply:
Note that we only delete state transitions for chunks that are present in the block, that takes care issues with missing chunks. It means that we already used state transitions data for those chunks, and since this block is final all further blocks can't be built on any block before that.

Reply:
But couldn't something like this happen?:

### Block at height 10:
Shards [0, 1, 2, 3] all have a new chunk
### Blocks at heights 11,12,13:
Have chunks for shards [0, 1, 2], but the chunk for shard 3 is missing in all blocks

Once we reach this point, block `#11` is final (3 confirmations) and `StateTransitionData` for block `#10` is deleted.

#### Block at height 14
Chunk producer produces a chunk on shard 3 and tries to produce a state witness for this chunk.
For that it needs to fetch `StateTransitionData` from block `#10`, shard 3, but it has been deleted.
All nodes deleted this data, so nobody is able to produce a chunk and this shard dies.

Reply:
Ah, but the iteration is only up to before `last_chunk_height_created`. Okaay, that takes care of the problem  :+1: